[BREAKING] Dvot 500/f/set account

jordipainan opened the pull request:
[BREAKING CHANGES]

This PR changes de behavior of setAccountInfo, refactors the mentioned tx to setAccount and adapts the required code for having all working properly.

Now with set account you can:
- Create an account only if a valid faucet package is provided
- Create an account without an infoURI
- Add/Del delegates in batch just changing the inner SetAccountTx TxType

The inner txtypes are maintained for backwards compatibility.

As a requirement for making the CI tests pass it also deletes the vocli
package and its tests, see commit message.

jordipainan committed:
vochain: make setAccount generic

jordipainan committed:
vochain: refactor setAccountInfo, create now requires faucet

jordipainan committed:
vochain: delete vocli

Since the numerous changes regarding account
creation and modification most of the vocli
tests need to be refactored.
This package is meant to be deleted, so
does not make sense to make the tests
work (for the second time).

New tests for the new CLI will be added
on the following commits.

jordipainan committed:
e2e: fix testTokenTxs tests

altergui reviewed (approved):
LGTM

jordipainan committed:
adapt code for optional attributes on SetAccounTx

p4u commented on api/wallet.go:
```diff
@@ -207,9 +207,9 @@ func (a *API) walletCreateHandler(msg *bearerstdapi.BearerStandardAPIdata, ctx *
+			SetAccount: &models.SetAccountTx{
+				Txtype:        models.TxType_SET_ACCOUNT_INFO_URI,
 				Nonce:         0,
 				InfoURI:       "none",
```
I understand this could be empty now.

p4u replied:
is this solved @jordipainan?

jordipainan replied:
Yes it CAN be empty, but this method needs to be refactored anyways for adding delegates too on the account bootstrap. I was planning to do it when tacking the client and the API with the tests.

I made it empty btw

jordipainan committed:
bump proto

jordipainan committed:
update gitignore

jordipainan committed:
rpcclient: unify create/set account

jordipainan committed:
add maxURLLength const for account infoURI

jordipainan committed:
dvotecli: add accounts on genesis generator

jordipainan committed:
vochaintest: use rpcclient SetAccount

jordipainan committed:
api_test: use CreateAccount tx on account creation

jordipainan committed:
vochain: add accounts parameter on NewGenesis()

jordipainan committed:
genesis: add CREATE_ACCOUNT tx type

jordipainan committed:
vochain: avoid possible nil pointers

jordipainan committed:
vochain state: avoid possible nil and update txcost

jordipainan committed:
vochain: refactor setAccount and Check/Deliver Tx values

This commit changes the logic for having all SetAccount
related tx using the same proto message (setAccountTx)
and different inner TxType depending on the tx to execute.
The inner TxType can be: CREATE_ACCOUNT,
SET_ACCOUNT_INFOURI, ADD_DELEGATE_FOR_ACCOUNT and
DEL_DELEGATE_FOR_ACCOUNT

It also changes how values for commiting to the state
are obtained. Currently most of the values that are
commited are taken from transformations made on
tx data received on the DeliverTx, which executes the
same logic as the CheckTx.
In order to have no data dependency and in order
to provide code clarity all values that are
used for committing to the state are no longer obtained
from the transformations made on the data obtained from
the tx on the DeliverTx phase, all the data
that is used for the commit phase is obtained from the
initial tx and is expected to be totally valid at this point
given that all the checks are made in the check/deliver tx
phase.
This does not mean that in the past the commit
phase could have incorrect data but after conversations
with the core team we found that using this approach
is better for code maintainability and avoids possible
security problems as the code grows.

Tests were adapted accordingly.

jordipainan committed:
vochain: add event listeners for set account & transfer tokens

p4u commented on api/wallet.go:
```diff
@@ -206,12 +206,13 @@ func (a *API) walletCreateHandler(msg *bearerstdapi.BearerStandardAPIdata, ctx *
 	}
 
 	stx := models.SignedTx{}
+	infoURI := string("none")
```
I'd let this just an empty string. There is already a zero-value for a string type so I don't see the need for setting an own one.

p4u commented on vochain/account.go:
```diff
@@ -67,9 +67,6 @@ func (a *Account) IsDelegate(addr common.Address) bool {
 
 // AddDelegate adds an address to the list of delegates for an account
 func (a *Account) AddDelegate(addr common.Address) error {
-	if a.IsDelegate(addr) {
```
Why this check is removed?

jordipainan replied:
Because is already checked on the on check and deliver tx, but I added it again as well on RemoveDelegate.

p4u commented on vochain/account.go:
```diff
@@ -208,251 +223,248 @@ func (v *State) IncrementAccountProcessIndex(accountAddress common.Address) erro
-		return fmt.Errorf("cannot create account %s: %w", accountAddress.String(), err)
+// ConsumeFaucetPayload consumes a given faucet payload storing
+// its key to the FaucetNonce tree so it can
+// only be used once
```
This line breaking does might not be necessary on the comment.

p4u commented on vochain/account.go:
```diff
@@ -208,251 +223,248 @@ func (v *State) IncrementAccountProcessIndex(accountAddress common.Address) erro
-	if acc != nil {
-		return ErrAccountAlreadyExists
+	// check from account
+	if from == types.EthereumZeroAddress {
```
What's the reason for this check? The from comes from a signature, so in theory the zero Address  will never be a from.

jordipainan replied:
Yup good one, removed.

p4u commented on vochain/account.go:
```diff
@@ -208,251 +223,248 @@ func (v *State) IncrementAccountProcessIndex(accountAddress common.Address) erro
+	binary.LittleEndian.PutUint64(b, faucetPayload.Identifier)
+	key := from.Bytes()
+	key = append(key, b...)
+	keyHash := crypto.Sha256(key)
```
Nit: you can do `keyHash := crypto.Sha256(append(from.Bytes(), b...))`

Is there a reason for using crypto.Sha256() here? Everywhere in the code we use ethereum.HashRaw() as default hashing function.

jordipainan replied:
Applied suggestion and I've also changed the `crypto.Sha256` for our own `ethereum.HashRaw` it does not make sense to use another hash function, here the autocompletion played bad :S

p4u commented on vochain/account.go:
```diff
@@ -208,251 +223,248 @@ func (v *State) IncrementAccountProcessIndex(accountAddress common.Address) erro
-	// check faucet payload
-	if faucetPayload == nil {
-		return fmt.Errorf("faucet payload is nil")
+// CreateAccountTxCheck is an abstraction of the ABCI CheckTx method for creating an account
```
This looks a wrong IA comment for this method.

jordipainan replied:
hahah, I made it by myself. Changed for looking less IA and more human.

p4u commented on vochain/account.go:
```diff
@@ -208,251 +223,248 @@ func (v *State) IncrementAccountProcessIndex(accountAddress common.Address) erro
+		delegatesMap := make(map[common.Address]bool)
+		for _, v := range tx.Delegates {
+			delegateAddress := common.BytesToAddress(v)
+			if delegateAddress == types.EthereumZeroAddress {
```
I'd not consider EthereumZeroAddress as an special address. Its just too difficult to find it.

p4u commented on vochain/account.go:
```diff
@@ -208,251 +223,248 @@ func (v *State) IncrementAccountProcessIndex(accountAddress common.Address) erro
-	if err := v.SetAccount(accToAddr, accTo); err != nil {
-		return err
+// CreateAccount creates an account
+func (v *State) CreateAccount(accountAddress common.Address,
```
We do not use this line breaking for method definition.

p4u commented on vochain/account.go:
```diff
@@ -208,251 +223,248 @@ func (v *State) IncrementAccountProcessIndex(accountAddress common.Address) erro
+	return v.createAccount(accountAddress, infoURI, delegates, 0)
+}
+
+func (v *State) createAccount(accountAddress common.Address,
```
Same

p4u commented on vochain/account.go:
```diff
@@ -208,251 +223,248 @@ func (v *State) IncrementAccountProcessIndex(accountAddress common.Address) erro
+		return fmt.Errorf("invalid transaction")
+	}
+	if tx.Nonce == nil || tx.InfoURI == nil {
+		return fmt.Errorf("missing signature and/or transaction")
```
This error message seems wrong

p4u commented on vochain/account.go:
```diff
@@ -208,251 +223,248 @@ func (v *State) IncrementAccountProcessIndex(accountAddress common.Address) erro
-	// get txSender account
-	txSender, err := state.GetAccount(returnValues.TxSender, false)
+	var txAccountAddress common.Address
+	if len(tx.Account) == 0 || bytes.Equal(tx.Account, types.EthereumZeroAddress.Bytes()) {
```
Same with zero address... But in the other side I don't understand the reason behind this check. Why not just take the txSenderAddress and ignore Account? Or the other option would be to enforce tx.Account to be equal to txSenderAddress.

jordipainan replied:
because it can be a delegate and the signature points to the delegate address

p4u commented on vochain/account.go:
```diff
@@ -208,251 +223,248 @@ func (v *State) IncrementAccountProcessIndex(accountAddress common.Address) erro
-	// check txSender balance
-	if txSender.Balance < cost {
-		return nil, fmt.Errorf("unauthorized: %s", ErrNotEnoughBalance)
+	if txAccountAccount == nil {
```
Isn't this check already performed?

jordipainan replied:
nope, first time

p4u commented on vochain/account.go:
```diff
@@ -208,251 +223,248 @@ func (v *State) IncrementAccountProcessIndex(accountAddress common.Address) erro
+	if txAccountAccount == nil {
+		return ErrAccountNotExist
+	}
+	if infoURI == txAccountAccount.InfoURI {
```
This check also seems to be repeated

p4u commented on vochain/account.go:
```diff
@@ -208,251 +223,248 @@ func (v *State) IncrementAccountProcessIndex(accountAddress common.Address) erro
-	if returnValues.TxSender != returnValues.Account {
-		if !acc.IsDelegate(returnValues.TxSender) {
-			return nil, fmt.Errorf("tx sender is not a delegate")
+	if txSenderAddress == txAccountAddress {
```
How can txSenderAddress and txAccountAddress be different? One is obtained from the other.

jordipainan replied:
if delegates can be different, the txSenderAddress is the address sending the tx, the txAccountAddress is the address from the Account field if the tx if not empty, otherwise is the same as txSenderAddress

p4u commented on vochain/account.go:
```diff
@@ -461,250 +473,253 @@ func (v *State) SetAccount(accountAddress common.Address, account *Account) erro
+		account.ProcessIndex,
 	)
+	for _, l := range v.eventListeners {
+		l.OnSetAccount(accountAddress.Bytes(), account)
```
I think passing a the account itself makes more sense than just the bytes of the address. But take care on not passing it as a pointer but as a value.

jordipainan replied:
already passing the account, but changed to pass a copy

p4u commented on vochain/account.go:
```diff
@@ -461,250 +473,253 @@ func (v *State) SetAccount(accountAddress common.Address, account *Account) erro
-	// check nonce
-	if tx.Nonce != treasurer.Nonce {
-		return common.Address{}, 0, fmt.Errorf("invalid nonce %d, expected: %d", tx.Nonce, treasurer.Nonce)
+	if len(tx.To) == 0 || bytes.Equal(tx.To, types.EthereumZeroAddress.Bytes()) {
```
What the zero address did to you?? :)

jordipainan replied:
nothing haha, is a basic check all the eth code does to prevent operating with the zero account, and coming from that world i took this habit.
Anyways, deleted all zeroaddr checks.

jordipainan replied:
update: still checking for delegates operation as the tx.Account is a field that can be different than txSender for those cases.

p4u commented on vochain/account.go:
```diff
@@ -461,250 +473,253 @@ func (v *State) SetAccount(accountAddress common.Address, account *Account) erro
-	// check delegate to add is not itself
-	if delAcc == *sigAddress {
-		return nil, fmt.Errorf("cannot add self to delegates list")
+	delegatesToSetMap := make(map[common.Address]bool)
```
I saw this code snipped already. Is it necessary to duplicate it? In theory the TxCheck functions must be called before modifying the sate, so one can relay on the other.

jordipainan replied:
they have different purposes, one checks duplicates against tx.delegates and the other against the account that will be eventually modified.

p4u commented on vochain/transaction.go:
```diff
@@ -185,54 +185,155 @@ func (app *BaseApplication) AddTx(vtx *VochainTx, commit bool) (*AddTxResponse,
+				if err != nil {
+					return nil, fmt.Errorf("createAccountTx: txSenderAddress %w", err)
+				}
+				delegates := make([][]byte, 0)
```
Third time I see this code

jordipainan replied:
moved this check out of the commit phase, and added a generic checkDuplicateDelegates util function

jordipainan committed:
apply changes requested

p4u commented on vochain/offchaindatahandler/offchaindatahandler.go:
```diff
@@ -149,7 +149,12 @@ func (c *OffChainDataHandler) OnRevealKeys(pid []byte, priv string, txindex int3
 	status models.ProcessStatus, txindex int32) {
 }
-
+func (c *OffChainDataHandler) OnSetAccount(addr []byte, account *vochain.Account) error {
```
AFAIK the address is found inside the account

jordipainan replied:
it is not, the address have never been part of the account struct

jordipainan replied:
@p4u Let me know if you are thinking about to change this because it has lots of implications on the code and I wont go ahead with this PR.

p4u replied:
You'r right, the address is the key for the merkle tree storing the Account values.

p4u commented on vochain/transaction.go:
```diff
@@ -185,54 +185,139 @@ func (app *BaseApplication) AddTx(vtx *VochainTx, commit bool) (*AddTxResponse,
-			return nil, fmt.Errorf("setAccountInfoTxCheck: %w", err)
+	case *models.Tx_SetAccount:
+		tx := vtx.Tx.GetSetAccount()
+		var err error
```
I think we don't need to declare this error var here

jordipainan replied:
its a minor i fixed on #622 but lets fix it here too :)

p4u commented on vochain/transaction.go:
```diff
@@ -185,54 +185,139 @@ func (app *BaseApplication) AddTx(vtx *VochainTx, commit bool) (*AddTxResponse,
+				); err != nil {
+					return nil, fmt.Errorf("setAccountDelegate: %w", err)
+				}
+			case models.TxType_DEL_DELEGATE_FOR_ACCOUNT:
```
Nit: let's add either a blank line between "case" statements or not (I think we should add them here).

jordipainan replied:
I guess you are saying this because of the next switch case, I've added a new line on those because is less clear when reading because you have more code, but in the former I do not think is necessary cause someone can clearly distinguish between each switch case.
Said so, I've added it as you are recommending.

jordipainan committed:
change genesis chainID and date
